File: src/lib/answerChunks.ts
// Chunk-level source attribution for ChatGPT/SearchGPT answers.
//
// A web-search answer is assembled from chunks, each ending with one or more
// markers like "... standout right now. [1]" whose numbers map 1:1 to
// links_attached[].position. Splitting on those markers gives every source
// the exact text it backs. Mirror of llmi_be/app/services/answer_chunks.py —
// keep the two implementations in sync.

export interface AnswerChunk {
  text: string;
  positions: number[];
  urls: string[];
}

const MARKER_RUN = /(?:\s*\[\d+\])+/g;
const MARKER_NUM = /\[(\d+)\]/g;

export function parseAnswerChunks(
  answerText: string | null | undefined,
  linksAttached: Array<{ url?: string; position?: number }> | null | undefined
): AnswerChunk[] {
  const text = answerText || '';
  const urlByPos = new Map<number, string>();
  (linksAttached || []).forEach((link, i) => {
    const pos = link.position ?? i + 1;
    if (!urlByPos.has(pos) && link.url) urlByPos.set(pos, link.url);
  });

  const chunks: AnswerChunk[] = [];
  let cursor = 0;
  for (const m of text.matchAll(MARKER_RUN)) {
    const chunkText = text.slice(cursor, m.index).trim();
    const positions = Array.from(m[0].matchAll(MARKER_NUM)).map(x => parseInt(x[1], 10));
    if (chunkText || positions.length > 0) {
      chunks.push({
        text: chunkText,
        positions,
        urls: positions.map(p => urlByPos.get(p)).filter((u): u is string => !!u),
      });
    }
    cursor = (m.index ?? 0) + m[0].length;
  }
  const tail = text.slice(cursor).trim();
  if (tail) chunks.push({ text: tail, positions: [], urls: [] });
  return chunks;
}

File: llmi_be/app/services/answer_chunks.py
"""Chunk-level source attribution for ChatGPT/SearchGPT answers.

A web-search answer is assembled from chunks, each ending with one or more
source markers like " ... standout right now. [1]" whose numbers map 1:1 to
links_attached[].position. Splitting the answer on those markers gives every
source the exact text it backs — the foundation for per-source brand
analysis (which competitor is mentioned in the chunk sourced from URL X).

Pure functions, no I/O. Mirrored on the frontend in src/lib/answerChunks.ts —
keep the two implementations in sync.
"""
import re

_MARKER_RUN = re.compile(r"(?:\s*\[\d+\])+")
_MARKER_NUM = re.compile(r"\[(\d+)\]")


def parse_answer_chunks(answer_text: str, links_attached: list) -> list:
    """Split an answer into source-attributed chunks.

    Returns [{"text", "positions": [int], "urls": [str]}]. A chunk is the text
    preceding a run of consecutive markers ("...text [1][2]") and is
    attributed to every marker in the run. Trailing text after the last
    marker (or a marker-less answer) yields a final chunk with no positions.
    """
    text = answer_text or ""
    url_by_pos = {}
    for i, link in enumerate(links_attached or []):
        pos = link.get("position") or (i + 1)
        if pos not in url_by_pos and link.get("url"):
            url_by_pos[pos] = link["url"]

    chunks = []
    cursor = 0
    for m in _MARKER_RUN.finditer(text):
        chunk_text = text[cursor:m.start()].strip()
        positions = [int(n) for n in _MARKER_NUM.findall(m.group(0))]
        if chunk_text or positions:
            chunks.append({
                "text": chunk_text,
                "positions": positions,
                "urls": [url_by_pos[p] for p in positions if p in url_by_pos],
            })
        cursor = m.end()
    tail = text[cursor:].strip()
    if tail:
        chunks.append({"text": tail, "positions": [], "urls": []})
    return chunks


def chunks_for_url(answer_text: str, links_attached: list, url: str) -> list:
    """Texts of all chunks attributed to the given source URL."""
    return [
        c["text"] for c in parse_answer_chunks(answer_text, links_attached)
        if url in c["urls"] and c["text"]
    ]
